File: src/main/services/deletion.ts
import { eden } from "@main/client";
import { chunk } from "es-toolkit";

export const DELETION_BATCH_SIZE = 500;

export type DeletionAccepted = {
    kind: "accepted";
    deletionJobId: string;
    status: "pending";
    deletionJobToken?: string;
};

export type DeletionCompleted = {
    kind: "completed";
    deletedCount: number;
};

export type DeletionResult = DeletionAccepted | DeletionCompleted;

export type BatchDeletionOutcome = {
    requestedIds: string[];
    acceptedIds: string[];
    jobs: DeletionAccepted[];
    errorMessage?: string;
};

export async function deleteDriveItems(uuids: string[]) {
    return runDeletionBatches(uuids, async (page) => {
        const { data, error } = await eden.akasha.content.delete_many.post({
            uuids: page,
        });
        return requireAccepted(resolveDeletionResult(data, error));
    });
}

export async function runDeletionBatches(
    ids: string[],
    request: (page: string[]) => Promise<DeletionAccepted | null>,
    batchSize = DELETION_BATCH_SIZE,
): Promise<BatchDeletionOutcome> {
    const requestedIds = [...new Set(ids)];
    if (requestedIds.length === 0) {
        return { requestedIds, acceptedIds: [], jobs: [] };
    }

    const acceptedIds: string[] = [];
    const jobs: DeletionAccepted[] = [];

    for (const page of chunk(requestedIds, batchSize)) {
        try {
            const job = await request(page);
            if (job) jobs.push(job);
            acceptedIds.push(...page);
        } catch (error) {
            return {
                requestedIds,
                acceptedIds,
                jobs,
                errorMessage: error instanceof Error ? error.message : String(error),
            };
        }
    }

    return { requestedIds, acceptedIds, jobs };
}

export function requireBatchAccepted(outcome: BatchDeletionOutcome) {
    if (outcome.acceptedIds.length === 0) {
        throw new Error(outcome.errorMessage || "delete_failed");
    }
    return outcome;
}

export function resolveDeletionResult(
    data: unknown,
    error: { status: number; value: unknown } | null,
): DeletionResult {
    if (error) {
        // Eden may surface explicit 202 responses on the error channel.
        if (error.status === 202) {
            const accepted = asDeletionAccepted(error.value);
            if (accepted) return accepted;
            const completed = asDeletionCompleted(error.value);
            if (completed) return completed;
        }
        throw new Error(toErrorMessage(error.value));
    }

    const accepted = asDeletionAccepted(data);
    if (accepted) return accepted;
    const completed = asDeletionCompleted(data);
    if (completed) return completed;

    throw new Error("unexpected_deletion_response");
}

/** Returns the accepted job, or null when the batch completed synchronously. */
function requireAccepted(result: DeletionResult): DeletionAccepted | null {
    if (result.kind === "accepted") return result;
    if (result.kind === "completed") return null;
    throw new Error("unexpected_deletion_response");
}

function asDeletionAccepted(value: unknown): DeletionAccepted | null {
    if (!value || typeof value !== "object") return null;
    if (!("deletionJobId" in value) || typeof value.deletionJobId !== "string") return null;
    if (!("status" in value) || value.status !== "pending") return null;
    return {
        kind: "accepted",
        deletionJobId: value.deletionJobId,
        status: "pending",
        deletionJobToken:
            "deletionJobToken" in value && typeof value.deletionJobToken === "string"
                ? value.deletionJobToken
                : undefined,
    };
}

function asDeletionCompleted(value: unknown): DeletionCompleted | null {
    if (!value || typeof value !== "object") return null;
    if (!("status" in value) || value.status !== "completed") return null;
    return {
        kind: "completed",
        deletedCount:
            "deletedCount" in value && typeof value.deletedCount === "number"
                ? value.deletedCount
                : 0,
    };
}

function toErrorMessage(value: unknown) {
    if (typeof value === "string") return value;
    if (value && typeof value === "object" && "message" in value) {
        return String(value.message);
    }
    return String(value);
}

File: internal/drive/deletion.go
package drive

import (
	"errors"
	"fmt"

	"github.com/samber/lo"
)

const DeletionBatchSize = 500

type DeletionAccepted struct {
	Kind             string `json:"kind"`
	DeletionJobID    string `json:"deletionJobId"`
	Status           string `json:"status"`
	DeletionJobToken string `json:"deletionJobToken,omitempty"`
}

type BatchDeletionOutcome struct {
	RequestedIDs []string           `json:"requestedIds"`
	AcceptedIDs  []string           `json:"acceptedIds"`
	Jobs         []DeletionAccepted `json:"jobs"`
	ErrorMessage string             `json:"errorMessage,omitempty"`
}

type deletionRequester func(page []string) (*DeletionAccepted, error)

func runDeletionBatches(ids []string, request deletionRequester, batchSize int) BatchDeletionOutcome {
	if batchSize <= 0 {
		batchSize = DeletionBatchSize
	}
	requested := uniqueStrings(ids)
	if len(requested) == 0 {
		return BatchDeletionOutcome{RequestedIDs: requested, AcceptedIDs: []string{}, Jobs: []DeletionAccepted{}}
	}

	accepted := make([]string, 0, len(requested))
	jobs := make([]DeletionAccepted, 0)
	for _, page := range chunkStrings(requested, batchSize) {
		job, err := request(page)
		if err != nil {
			return BatchDeletionOutcome{
				RequestedIDs: requested,
				AcceptedIDs:  accepted,
				Jobs:         jobs,
				ErrorMessage: err.Error(),
			}
		}
		if job != nil {
			jobs = append(jobs, *job)
		}
		accepted = append(accepted, page...)
	}
	return BatchDeletionOutcome{RequestedIDs: requested, AcceptedIDs: accepted, Jobs: jobs}
}

func requireBatchAccepted(outcome BatchDeletionOutcome) (BatchDeletionOutcome, error) {
	if len(outcome.AcceptedIDs) == 0 {
		msg := outcome.ErrorMessage
		if msg == "" {
			msg = "delete_failed"
		}
		return outcome, errors.New(msg)
	}
	return outcome, nil
}

type edenError struct {
	Status int
	Value  any
}

// resolveDeletionJob returns the job the server accepted for one page of a batch
// delete. A nil job with a nil error means the page was already deleted, so there
// is nothing left to schedule.
func resolveDeletionJob(data any, err *edenError) (*DeletionAccepted, error) {
	if err != nil {
		if err.Status == 202 {
			if accepted := asDeletionAccepted(err.Value); accepted != nil {
				return accepted, nil
			}
			if isDeletionCompleted(err.Value) {
				return nil, nil
			}
		}
		return nil, errors.New(toErrorMessage(err.Value))
	}

	if accepted := asDeletionAccepted(data); accepted != nil {
		return accepted, nil
	}
	if isDeletionCompleted(data) {
		return nil, nil
	}
	return nil, errors.New("unexpected_deletion_response")
}

func asDeletionAccepted(value any) *DeletionAccepted {
	record, ok := asRecord(value)
	if !ok {
		return nil
	}
	jobID, _ := record["deletionJobId"].(string)
	if jobID == "" {
		return nil
	}
	status, _ := record["status"].(string)
	if status != "pending" {
		return nil
	}
	out := &DeletionAccepted{
		Kind:          "accepted",
		DeletionJobID: jobID,
		Status:        "pending",
	}
	if token, ok := record["deletionJobToken"].(string); ok {
		out.DeletionJobToken = token
	}
	return out
}

func isDeletionCompleted(value any) bool {
	record, ok := asRecord(value)
	if !ok {
		return false
	}
	status, _ := record["status"].(string)
	return status == "completed"
}

func uniqueStrings(ids []string) []string {
	return lo.Uniq(ids)
}

func chunkStrings(ids []string, size int) [][]string {
	if size <= 0 {
		return [][]string{ids}
	}
	return lo.Chunk(ids, size)
}

func formatPartialDeleteLog(outcome BatchDeletionOutcome, action string) map[string]any {
	return map[string]any{
		"channel":     "drive:delete:items",
		"action":      action,
		"stage":       "delete-items-partial",
		"ids":         outcome.RequestedIDs,
		"acceptedIds": outcome.AcceptedIDs,
		"failedCount": len(outcome.RequestedIDs) - len(outcome.AcceptedIDs),
		"jobCount":    len(outcome.Jobs),
		"error":       outcome.ErrorMessage,
	}
}

func invalidActionError() error {
	return fmt.Errorf("INVALID_ACTION")
}
